src/internal/pjs/api_server.go

```diff
@@ -189,3 +176,108 @@ func (a *apiServer) ProcessQueue(srv pjs.API_ProcessQueueServer) (retErr error)
 		}
 	}
 }
+
+func (a *apiServer) WalkJob(req *pjs.WalkJobRequest, srv pjs.API_WalkJobServer) (err error) {
+	ctx := pctx.Child(srv.Context(), "walkJob")
+
+	// handle job context and request validation.
+	var id pjsdb.JobID
+	if req.Context != "" {
+		id, err = a.resolveJobCtx(ctx, req.Context)
+		if err != nil {
+			return errors.Wrap(err, "walk job")
+		}
+	} else {
+		// TODO(PJS): do auth.
+	}
+	// TODO(PJS): remove this once auth is implemented.
+	reqID := pjsdb.JobID(req.Job.Id)
+	if id != reqID {
+		log.Error(ctx, "job context token does not match requested job.", zap.Int64("request.Job.ID", req.Job.Id))
+		id = reqID // defaulting to this for testing until auth is implemented.
+	}
+
+	// walk and stream back results.
+	var jobs []pjsdb.Job
+	if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+		jobs, err = pjsdb.WalkJob(ctx, sqlTx, id, pjsdb.WalkAlgorithm(req.Algorithm.Number()))
+		return errors.Wrap(err, "with tx")
+	}); err != nil {
+		return errors.Wrap(err, "walk job")
+	}
+	for _, job := range jobs {
+		jobInfo, err := toJobInfo(job)
+		if err != nil {
+			return errors.Wrap(err, "walk job")
+		}
+		resp := &pjs.ListJobResponse{
+			Id:   jobInfo.Job,
+			Info: jobInfo,
+			Details: &pjs.JobInfoDetails{
+				JobInfo: jobInfo,
+			},
+		}
+		if err := srv.Send(resp); err != nil {
+			return errors.Wrap(err, "walk job")
```

Comment:
wrap with `send`.  The caller of this function should wrap the error with `WalkJob`, not this function itself.  Only wrap errors with information the caller can't know.

In this case, you should Wrapf with `job` here.  That way, when Send fails we know how deep into `jobs` we are.